### ISSUE TITLE
Custom variable/function scope highlighting

### ISSUE THREAD

**Issue by asheeshr** [facelessuser/BracketHighlighter#545]
##Description

Hello. I mainly program in C and was wondering if the package could support custom rules that could highlight constructor/destructor calls. For example, code in C would have this pattern:

```
struct my_struct variable = my_struct_create();
...

...
my_struct_destroy(variable);
```

I would like to highlight the create and destroy call pair if it exists in the same scope or show an error/warning if there isn't a pair found. Is that something that would be possible via this package?

I saw that the C style highlighting example in the settings file and was trying something like the following in the user settings file but wasn't sure how to make the open/close conditions look for the same variable and how to get this to trigger.

```
    "user_scope_brackets": [    {
            "name": "c_ctor_dtor",
            "open": "*(_create\\()",
            "close": "*(_destroy\\()",
            "style": "c_define",
            "scope_exclude": ["string", "comment"],
            "language_filter": "whitelist",
            "language_list": ["C", "C99", "C Improved" ],
            // "sub_bracket_search": "true",
            // "enabled": true
        }
]
```

To summarize, is the create/destroy pattern for variables something this package can support? How would I go about getting the rule to work for this use case?

**Comment by facelessuser** [comment]
You are going to have conflicts with the normal round bracket rule with the way you are defining this. I would just use a look ahead to ensure the keyword precedes a `(`. Also starting a pattern with `*` is a bad idea. I think you meant `.*`. You don't have it in the capture group, so I assume you don't need it, and since it is looking for 0 or 1, I don't think it matters:

```
    "user_brackets": [
        {
            "name": "c_ctor_dtor",
            "open": "(_create)(?=\\()",
            "close": "(_destroy)(?=\\()",
            "style": "c_define",
            "scope_exclude": ["string", "comment"],
            "language_filter": "whitelist",
            "language_list": ["C", "C99", "C Improved" ],
            // "sub_bracket_search": "true",
            "enabled": true
        },
```

Now, if you wanted to capture `my_struct_create` instead of just `_create`, and then ensure the start name matches the closing name, then you may want to capture the `.*` and then use a bh_plugin to compare the start and end name to confirm they match.

Now, BH expects brackes to follow a logical flow `{ ( [ ] ) }` not `{ ( [ ) } ]` When it comes to structure creation and destruction, they won't always follow the first pattern as you may destroy them in other orders, and they may be nested under other kinds of brackets conditionally which will break BH's ability to match the correct bracket. This happens sometimes if complex `#if` preprocessors. Preprocessors can really cause a problem with BH as they are evaluated at a different time agnostic of other brackets, and I feel your allocation and deallocation of structures will kind of fall into this same category.

Anyways, hopefully that info helps.

@gir-bot remove S: triage
@gir-bot add T: support
